### PR TITLE
feat: mutable storage ISMs

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Some chains like zkSync do not support eip1167 (minimal/meta) proxies. This PR adds an alternative storage based multisig and aggregation ISM for use on these chains.

### Drive-by changes

Simplify CLI multisig interactive config builder. Remove stale multisig config.

### Related issues

None

### Backward compatibility

Yes, relayer already supports this module type

### Testing

Contract unit tests
Manual CLI tests

Screenshot 2024-10-02 at 4 05 08 PM

### CODE REVIEW

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts [hunk 1]
```diff
@@ -220,18 +224,55 @@ export class HyperlaneIsmFactory extends HyperlaneApp<ProxyFactoryFactories> {
...
-        : this.getContracts(destination).staticMessageIdMultisigIsmFactory;
 
-    const address = await this.deployStaticAddressSet(
-      destination,
-      multisigIsmFactory,
-      config.validators,
-      logger,
-      config.threshold,
-    );
+    const deployStatic = (factory: StaticThresholdAddressSetFactory) =>
```
Comment (paulbalaji) [inline typescript/sdk/src/ism/HyperlaneIsmFactory.ts:228]: will need to dupe this change to `EvmIsmModule` too https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/471002bd1a3c1d36e26bf7f54c5d0bed70bf1543/typescript/sdk/src/ism/EvmIsmModule.ts#L433

File: typescript/cli/src/config/ism.ts [hunk 2]
```diff
@@ -106,68 +111,51 @@ export async function createAdvancedIsmConfig(
...
-    const threshold = parseInt(thresholdInput, 10);
+export const createMultisigConfig = async (
+  ismType: MultisigIsmConfig['type'],
+): Promise<MultisigIsmConfig> => {
+  const validatorsInput = await input({
+    message:
+      'Enter validator addresses (comma separated list) for multisig ISM:',
+  });
+  const validators = validatorsInput.split(',').map((v) => v.trim());
+  const threshold = parseInt(
```
Comment (paulbalaji) [inline typescript/cli/src/config/ism.ts:137]: maybe also check the given threshold >=1 and <= validators.length

should we also recommend a value?

File: typescript/sdk/src/ism/HyperlaneIsmFactory.ts (hunk 1 above)
Comment (yorhodes) [reply typescript/sdk/src/ism/HyperlaneIsmFactory.ts:228]: brutal, I thought these were kinda using the same codepaths 😱

File: typescript/cli/src/config/ism.ts (hunk 2 above)
Comment (yorhodes) [reply typescript/cli/src/config/ism.ts:137]: trying to leave input validation to the zod schema
can make this change there I suppose though 0/n as a noop ISM could be useful

File: solidity/test/isms/MultisigIsm.t.sol [hunk 3]
```diff
@@ -11,13 +11,16 @@ import {TestMailbox} from "../../contracts/test/TestMailbox.sol";
...
+import {IThresholdAddressFactory} from "../../contracts/interfaces/IThresholdAddressFactory.sol";
 import {TypeCasts} from "../../contracts/libs/TypeCasts.sol";
 import {MerkleTreeHook} from "../../contracts/hooks/MerkleTreeHook.sol";
 import {TestMerkleTreeHook} from "../../contracts/test/TestMerkleTreeHook.sol";
 import {TestPostDispatchHook} from "../../contracts/test/TestPostDispatchHook.sol";
 import {Message} from "../../contracts/libs/Message.sol";
 import {ThresholdTestUtils} from "./IsmTestUtils.sol";
+import {StorageMessageIdMultisigIsm, StorageMerkleRootMultisigIsm, StorageMessageIdMultisigIsmFactory, StorageMerkleRootMultisigIsmFactory, AbstractStorageMultisigIsm} from "../../contracts/isms/multisig/StorageMultisigIsm.sol";
+
+uint8 constant MAX_VALIDATORS = 20;
```
Comment (nambrot) [inline solidity/test/isms/MultisigIsm.t.sol:23]: A comment on why this value is 20 might be nice

File: typescript/cli/src/tests/multisig.test.ts [hunk 4]
```diff
@@ -1,41 +0,0 @@
-import { expect } from 'chai';
```
Comment (nambrot) [inline typescript/cli/src/tests/multisig.test.ts:1]: Why is this getting removed?

**Review by nambrot — COMMENTED** [review]
I mostly did a cursory review, so would like @aroralanuk to take a deeper look for the approval

File: solidity/contracts/isms/multisig/AbstractMultisigIsm.sol [hunk 5]
```diff
@@ -67,7 +67,7 @@
  * @notice Manages per-domain m-of-n Validator sets of AbstractMultisig that are used to verify
  * interchain messages.
  */
-abstract contract AbstractMultisigIsm is AbstractMultisig {
+abstract contract AbstractMultisigIsm is AbstractMultisig, IMultisigIsm {
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply solidity/contracts/isms/multisig/AbstractMultisigIsm.sol:71]: yeah idk why it thinks these are the same name, they are not

File: solidity/test/isms/MultisigIsm.t.sol (hunk 3 above)
Comment (yorhodes) [reply solidity/test/isms/MultisigIsm.t.sol:23]: this was pretty arbitrary tbh

Comment (yorhodes) [reply solidity/test/isms/MultisigIsm.t.sol:23]: bumped it from 10 to 20

File: typescript/cli/src/tests/multisig.test.ts (hunk 4 above)
Comment (yorhodes) [reply typescript/cli/src/tests/multisig.test.ts:1]: we no longer use this config schema

**Review by aroralanuk — APPROVED** [review]
Contracts look good

**Review by paulbalaji — APPROVED** [review]

**Comment by yorhodes** [comment]
Copying audit feedback over
> 1. It is recommended to use Ownable2stepUpgradeable instead of OwnableUpgradeable for StorageAggregationIsm and StorageMultisigIsm, as it can prevent transferring ownership to an incorrect address.
> 2. It is recommended to emit an event in setModulesAndThreshold() for the values before the change.

**Pull request merged**